astropy/config/paths.py

```diff
@@ -30,16 +30,15 @@
 ]
 
 
-def _get_dir_path(
-    rootname: str, cls: type[_SetTempPath], fallback: Literal["cache", "config"]
-) -> Path:
+def _get_dir_path(rootname: str, cls: type[_SetTempPath]) -> Path:
     # If using set_temp_x, that overrides all
     if (xch := cls._temp_path) is not None:
         path = xch / rootname
         if not path.is_file():
             path.mkdir(exist_ok=True)
         return path.resolve()
 
+    _, _, fallback = cls.__name__.rpartition("_")
     if (
         (xdg_dir := os.getenv(f"XDG_{fallback.upper()}_HOME")) is not None
```

Comment:
`fallback` is used for constructing the name of an environment variable. Having a `fallback` parameter with the type `Literal["cache", "config"]` lets type checkers ensure that the environment variable that will be looked up is valid. The change here removes that information from the type system because the `__name__` of a subclass of `_SetTempPath` is not restricted.

The current annotation is not perfect because it does not prevent calls like `_get_dir_path(..., set_temp_cache, "config")`, so it can be a good idea to introduce a class attribute for `set_temp_cache` and `set_temp_config` to eliminate such calling errors, but trying to extract something from the class name is not good enough.

Reply:
I hear you. How about this ?